advcpmv-0.8-8.32.patch

```diff
@@ -429,57 +527,61 @@ diff -aur coreutils-8.32/src/cp.c coreutils-8.32-patched/src/cp.c
 +
 +        /* how many files are we copying */
 +        char command[1024];
-+        sprintf( command, "find \"%s\" -type f | wc -l", file[0]);
++        char * _cEscapedString = escape_double_quotes( file[0] );
++        sprintf( command, "find \"%s\" -type f | wc -l", _cEscapedString );
```

Comment:
in case we are deadset on using popen, then don't use double quotes. Double quotes will have problem with filenames that has `$` in it. Use single quote and replace all single quote -> `'\''` instead.

Reply:
I'm not set on popen. It's what was used in the original source. Your example looks very nice, I'll test it out.

Reply:
How would your example above have to be adapted to handle the pipe to wc -l that is needed for the find file command and counting the number of files to copy/move it found?

Reply:
Hmm... At least for me, the single quote approach does not work if the file/directory contains single quotes or backticks and it doesn't work if the name contains a `$`. 

The find command would have to look like this:
```
sprintf( command, "find \'%s\' -type f | wc -l", _cEscapedString );
sprintf( command, "du -s \'%s\'", _cEscapedString );
```
and in the escape_double_quotes function it would have to look like this
```
if(_cUnescapedString[rpos] == '\'')
```
correct?

Reply:
> How would your example above have to be adapted to handle the pipe to wc -l

No need. Just count newlines by using fgets or fread in a loop.

Reply:
Okay, but how large needs the return buffer to be? With the piped wc command we will get back a buffer with a relatively fixed length. Without it, the size depends on the number of files to copy and the full path length of each file? We could set the value quite high, but it could still break depending on the users copy/move operation. Or am I thinking in a wrong direction?

Reply:
Return buffer is not an issue here at all. You could also read one char at a time with `fgetc`.

Though, I'd probably use `getline` since it doesn't have issue with larger lines.

Reply:
Okay. I'll check it out. Thank you for patiently helping. :+1:

Reply:
Having it look like this, works but throws a valgrind error.
```
        /* how many files are we copying */
        FILE *fp ;
        char output[1024];
        char fcmd[] = "find";
        fp = spawn(fcmd, (char *[]){ fcmd, file[0], NULL, "-type f", NULL });
        if ( fp == NULL)
            printf("failed to run find\r");
        else
        {
          char *line_buf = NULL;
          size_t line_buf_size = 0;
          int line_count = 0;
          ssize_t line_size = 0;
          line_size = getline(&line_buf, &line_buf_size, fp);
          while (line_size >= 0)
          {
              line_count++;
              line_size = getline(&line_buf, &line_buf_size, fp);
          }
          if ( line_count > n_files )
            g_iTotalFiles = line_count;
        }
        /* close */
        pclose(fp);
```
Error:
```
==27831== 120 bytes in 1 blocks are definitely lost in loss record 4 of 6
==27831==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgprel>
==27831==    by 0x49186C3: getdelim (iogetdelim.c:62)
==27831==    by 0x10EF79: getline (stdio.h:120)
==27831==    by 0x10EF79: do_copy (cp.c:689)
==27831==    by 0x10D87E: main (cp.c:1381)
...
==27831== LEAK SUMMARY:
==27831==    definitely lost: 120 bytes in 1 blocks
```
So sadly my usage of getline seems to be problematic. Do you have an idea how to fix it?

Reply:
I think

    while (line_size >= 0)

should be

    while (line_size > 0)

Reply:
or, do you expect blank lines?

And how does it handle files with newline in the name?

Reply:
Please follow the programmatic way to detect the number of files rather than spawning a command. Please refer to:

https://github.com/jarun/nnn/blob/23b54ba0cf968d7c3c5973642186223c70aa9cf6/src/nnn.c#L5580

Reply:
I don't know if find could output blank lines and what happens if there are newlines in the filenames. Did the original code handle them? Changing to >0 works the same for my testcases, but does not change the valgrind error.

Reply:
The example I shared will handle all cases, and is the correct way to do it. Simplified, it will only be a few lines.

When you say "improvement", it doesn't matter if the code original code worked in a scenario or not.

Reply:
> So sadly my usage of getline seems to be problematic. Do you have an idea how to fix it?

The spawn arguments were not correct. It should've been:

```c
        fp = spawn(fcmd, (char *[]){ fcmd, file[0], "-type", "f", NULL });
```

Also the valgrind error is due to not freeing `line_buf`, `getline` allocates the buffer so it needed to be freed once you're done.

```c
          char *line_buf = NULL;
          size_t line_buf_size = 0;
          int line_count = 0;
          while (getline(&line_buf, &line_buf_size, fp) > 0)
            ++line_count;
          free(line_buf);
          if ( line_count > n_files )
            g_iTotalFiles = line_count;
```

Reply:
@SlrG I think you'll also need to escape the escape char, `\`.
https://mywiki.wooledge.org/Quotes#Types_of_quoting

Reply:
Please update for completeness.